File: backend/models/resume.py
from pydantic import BaseModel, Field
from typing import Optional, List


class PersonalInfo(BaseModel):
    full_name: str = ""
    email: str = ""
    phone: str = ""
    location: str = ""
    linkedin: str = ""
    website: str = ""
    title: str = ""


class Experience(BaseModel):
    company: str = ""
    role: str = ""
    start_date: str = ""
    end_date: str = ""
    description: str = ""


class Education(BaseModel):
    institution: str = ""
    degree: str = ""
    field: str = ""
    year: str = ""
    gpa: str = ""


class Project(BaseModel):
    name: str = ""
    description: str = ""
    tech_stack: str = ""
    link: str = ""


class Certification(BaseModel):
    name: str = ""
    issuer: str = ""
    year: str = ""


class ResumeData(BaseModel):
    personal_info: Optional[PersonalInfo] = None
    summary: str = ""
    skills: List[str] = []
    experience: List[Experience] = []
    education: List[Education] = []
    projects: List[Project] = []
    certifications: List[Certification] = []


class GenerateResumeRequest(BaseModel):
    raw_input: str = Field(..., description="User's raw text or voice-to-text input")
    job_role: Optional[str] = Field(default="", description="Target job role for optimization")
    template: Optional[str] = Field(default="minimal")


class GenerateResumeResponse(BaseModel):
    resume_data: ResumeData
    ats_score: int
    suggestions: List[str]


class ImproveSectionRequest(BaseModel):
    section: str = Field(..., description="Section name: summary, experience, skills, etc.")
    content: str = Field(..., description="Current content to improve")
    job_role: Optional[str] = ""


class ImproveSectionResponse(BaseModel):
    improved_content: str
    suggestions: List[str]
    keywords: List[str]

File: frontend/src/types/resume.ts
export interface PersonalInfo {
  full_name: string;
  email: string;
  phone: string;
  location: string;
  linkedin: string;
  website: string;
  title: string;
}

export interface Experience {
  company: string;
  role: string;
  start_date: string;
  end_date: string;
  description: string;
}

export interface Education {
  institution: string;
  degree: string;
  field: string;
  year: string;
  gpa: string;
}

export interface Project {
  name: string;
  description: string;
  tech_stack: string;
  link: string;
}

export interface Certification {
  name: string;
  issuer: string;
  year: string;
}

export interface Resume {
  personal_info: PersonalInfo;
  summary: string;
  skills: string[];
  experience: Experience[];
  education: Education[];
  projects: Project[];
  certifications: Certification[];
}

export type TemplateType = 'minimal' | 'corporate' | 'creative';

export interface SavedResume {
  id: number;
  name: string;
  resume: Resume;
  template: TemplateType;
  createdAt: string;
}

export interface ResumeState {
  resume: Resume;
  template: TemplateType;
  atsScore: number;
  suggestions: string[];
  isGenerating: boolean;
  isImproving: boolean;
  savedResumes: SavedResume[];
}

export type ResumeAction =
  | { type: 'SET_RESUME'; payload: Resume }
  | { type: 'UPDATE_FIELD'; field: keyof Resume; value: unknown }
  | { type: 'UPDATE_PERSONAL'; payload: Partial<PersonalInfo> }
  | { type: 'SET_TEMPLATE'; payload: TemplateType }
  | { type: 'SET_ATS'; score: number; suggestions: string[] }
  | { type: 'SET_GENERATING'; payload: boolean }
  | { type: 'SET_IMPROVING'; payload: boolean }
  | { type: 'SAVE_RESUME'; name: string }
  | { type: 'LOAD_RESUME'; payload: { resume: Resume; template: TemplateType } }
  | { type: 'DELETE_RESUME'; id: number }
  | { type: 'RESET' };

export interface ResumeContextValue extends ResumeState {
  setResume: (data: Resume) => void;
  updateField: (field: keyof Resume, value: unknown) => void;
  updatePersonal: (data: Partial<PersonalInfo>) => void;
  setTemplate: (template: TemplateType) => void;
  setAts: (score: number, suggestions: string[]) => void;
  setGenerating: (v: boolean) => void;
  setImproving: (v: boolean) => void;
  saveResume: (name: string) => void;
  loadResume: (data: { resume: Resume; template: TemplateType }) => void;
  deleteResume: (id: number) => void;
  reset: () => void;
}
